core/src/main/scala/org/apache/spark/scheduler/TaskSetBlacklist.scala

```diff
@@ -0,0 +1,136 @@
+   * executors on a node even after they have died.
+   */
+  private val nodeToExecsWithFailures: HashMap[String, HashSet[String]] = new HashMap()
+  private val nodeToBlacklistedTasks: HashMap[String, HashSet[Int]] = new HashMap()
+  private val blacklistedExecs: HashSet[String] = new HashSet()
+  private val blacklistedNodes: HashSet[String] = new HashSet()
+
+  /**
+   * Return true if this executor is blacklisted for the given task.  This does *not*
+   * need to return true if the executor is blacklisted for the entire stage.
+   * That is to keep this method as fast as possible in the inner-loop of the
+   * scheduler, where those filters will have already been applied.
+   */
+  def isExecutorBlacklistedForTask(
+      executorId: String,
+      index: Int): Boolean = {
+    execToFailures.get(executorId)
+      .map { execFailures =>
+        val count = execFailures.taskToFailureCountAndExpiryTime.get(index).map(_._1).getOrElse(0)
+        count >= MAX_TASK_ATTEMPTS_PER_EXECUTOR
+      }
+      .getOrElse(false)
+  }
+
+  def isNodeBlacklistedForTask(
+      node: String,
+      index: Int): Boolean = {
+    nodeToBlacklistedTasks.get(node)
+      .map(_.contains(index))
+      .getOrElse(false)
+  }
+
+  /**
+   * Return true if this executor is blacklisted for the given stage.  Completely ignores whether
+   * anything to do with the node the executor is on.  That
+   * is to keep this method as fast as possible in the inner-loop of the scheduler, where those
+   * filters will already have been applied.
+   */
+  def isExecutorBlacklistedForTaskSet(executorId: String): Boolean = {
+    blacklistedExecs.contains(executorId)
+  }
+
+  def isNodeBlacklistedForTaskSet(node: String): Boolean = {
+    blacklistedNodes.contains(node)
+  }
+
+  private[scheduler] def updateBlacklistForFailedTask(
+      host: String,
+      exec: String,
+      index: Int): Unit = {
+    val execFailures = execToFailures.getOrElseUpdate(exec, new ExecutorFailuresInTaskSet(host))
+    execFailures.updateWithFailure(index, clock.getTimeMillis() + TIMEOUT_MILLIS)
+
+    // check if this task has also failed on other executors on the same host -- if its gone
+    // over the limit, blacklist it from the entire host
+    val execsWithFailuresOnNode = nodeToExecsWithFailures.getOrElseUpdate(host, new HashSet())
+    execsWithFailuresOnNode += exec
+    val failuresOnHost = execsWithFailuresOnNode.toIterator.flatMap { exec =>
+      execToFailures.get(exec).map { failures =>
+        // We count task attempts here, not the number of unique executors with failures.  This is
+        // because jobs are aborted based on the number task attempts; if we counted unique
+        // executors, it would be hard to config to ensure that you try another
+        // node before hitting the max number of task failures.
+        failures.taskToFailureCountAndExpiryTime.getOrElse(index, (0, 0))._1
```

Comment:
I am not sure if I am missing it, but where are we cleaning up expired failures ? (Note that the original impl I had never cleaned up the map).
Shouldn't executor failure expiry not evict it from the node failure list as well (implicitly) ?

I mean - 
ex1_n1 failed, blacklisted for 10s (say)
ex2_n1 failed after 20s - we should not consider ex1_n1 failure as relevant now for n1 ?